### PR TITLE
[Spec Decode][Perf] Optimize DSpark Markov head with addmm

### PR DESCRIPTION
**Pull request by yzyDavid**
## Purpose

DSpark produces the base logits for all speculative positions in one LM-head
call, then applies the dense Markov projection sequentially for each draft
step. The existing path launches a projection and a separate elementwise add
for every step.

This PR makes the DSpark sample layout step-major, so each step owns a
contiguous `[request, vocab]` base-logits slice, and adds the Markov bias with:

```python
base_logits.addmm_(markov_embed, weight.t())
```

The beta=1 GEMM writes directly into the LM-head output after its final read.
This removes the separate add kernel without allocating an out-of-place
result. The fast path is limited to NVIDIA CUDA, an unquantized replicated
Markov head, compatible dtypes, contiguous logits, unit logit scale, and no
soft cap. All other configurations keep the existing non-mutating path.

The fixed `max_num_reqs` step stride also keeps the layout valid when full CUDA
Graphs pad the request batch. DFlash remains request-major.

### Duplicate-work check

I searched open vLLM PRs for DSpark, Markov head, addmm, in-place, and
step-major changes. I did not find another PR implementing this dense
step-major/in-place path.

- #49969 is an opt-in top-k Markov projection that restricts the candidate
  rows. This PR keeps the exact dense projection and optimizes its memory and
  kernel sequence. The approaches are conceptually complementary, although
  they touch two of the same files and may need a rebase if both land.
- #50424 adds quantization plumbing to the Markov head; this PR deliberately
  falls back for quantized heads.
- #47584 adds an opt-in FP8 draft LM head, not a Markov-bias addmm path.

No documentation or public configuration changes are required.

## Test Plan

1. Run focused Markov-head tests covering replication, in-place pointer and
   numerical equivalence, and soft-cap/scale/non-contiguous fallbacks.
2. Run CUDA layout tests covering anchor DSpark, fill-in DSpark, DFlash, and
   full-graph request padding.
3. Compare the original loop, out-of-place `torch.addmm`, and in-place
   `addmm_` with eager and CUDA Graph microbenchmarks on H20, BF16 and FP16,
   batch sizes 1 through 64.
4. Inspect an H20 CUDA timeline to verify the kernel and memcpy sequence.
5. Serve Qwen3-8B with `dspark_qwen3_8b_block7`, TP=1, BF16, default
   `torch.compile`, and FULL_AND_PIECEWISE CUDA Graphs. Compare upstream main
   and this PR with repeated greedy serving benchmarks, fixed outputs, and the
   full GSM8K test split.

Focused commands, run inside the project runtime container:

```bash
pytest -q --confcutdir=tests/models \
  tests/models/test_dspark_mla.py::test_dspark_markov_head_is_replicated \
  tests/models/test_dspark_mla.py::test_dspark_markov_head_fuses_projection_and_bias_add_in_place \
  tests/models/test_dspark_mla.py::test_dspark_markov_head_falls_back_when_inplace_is_not_safe

pytest -q --confcutdir=tests/v1/spec_decode \
  tests/v1/spec_decode/test_dspark_sample_layout.py
```

Ruff check, Ruff format check, `py_compile`, and `git diff --check` were also
run on the changed Python sources after rebasing onto main.

### Focused correctness

- Markov-head tests: 5 passed.
- DSpark/DFlash CUDA layout tests: 3 passed.
- All 14 BF16/FP16 microbenchmark cases had 100% final-token agreement.
- Fixed real-checkpoint greedy outputs: main and this PR were exactly equal on
  all 16 prompts.
- Both main and this PR completed backbone/DSpark-head compilation and
  PIECEWISE, FULL, and DSpark CUDA Graph capture.

### H20 microbenchmark

The CUDA Graph benchmark used 20 warmups, 51 samples, and 10 replays per
sample. Across batch sizes 1, 2, 4, 8, 16, 32, and 64:

- BF16 in-place versus original: 4.68% to 35.65% lower latency.
- FP16 in-place versus original: 4.74% to 35.64% lower latency.
- In-place versus out-of-place addmm: 4.16% to 13.56% lower latency.

At BF16 batch 16, the seven-step timeline changed from:

```text
7 x (embedding -> GEMM beta=0 -> add -> argmax)
```

to:

```text
7 x (embedding -> GEMM beta=1 -> argmax)
```

The out-of-place addmm trace contained seven D2D initialization copies. The
in-place trace contained zero.

### Repeated checkpoint E2E

`vllm bench serve` used temperature 0, seed 0, 128 input tokens, 128 output
tokens, and ignored EOS. Each arm ran three repetitions. Concurrency 1 used 64
prompts per repetition; the other levels used 256. All requests completed with
zero failures.

| Concurrency | Main output tok/s | This PR output tok/s | Delta | Main acceptance | This PR acceptance |
|---:|---:|---:|---:|---:|---:|
| 1 | 350.09 +/- 10.01 | 363.46 +/- 0.21 | +3.82% | 26.54% | 26.17% |
| 8 | 1658.19 +/- 108.51 | 1732.38 +/- 6.01 | +4.47% | 25.41% | 25.39% |
| 32 | 1982.21 +/- 9.46 | 1989.71 +/- 16.44 | +0.38% | 25.38% | 25.51% |
| 64 | 2052.04 +/- 35.42 | 2087.94 +/- 10.09 | +1.75% | 25.16% | 25.45% |

Values are mean +/- sample standard deviation. The concurrency-32 difference
is within run-to-run noise and should be read as neutral. Acceptance stayed
aligned, with a maximum absolute difference of 0.37 percentage points.

### GSM8K

All 1,319 test examples used five-shot prompting, temperature 0, at most 256
new tokens, and concurrency 64. There were no invalid answers.

| Configuration | Correct | Accuracy |
|---|---:|---:|
| Target only | 1170 / 1319 | 88.704% |
| Main DSpark | 1168 / 1319 | 88.552% |
| This PR DSpark | 1171 / 1319 | 88.779% |

There is no model-quality regression signal.

### AI-assistance disclosure

This change was developed with OpenAI Codex assistance. I reviewed every
changed line, understand the implementation and validation, and accept
responsibility for explaining and maintaining the contribution.

---
<details>
<summary>Essential Elements of an Effective PR Description Checklist</summary>

- [x] The purpose of the PR is described.
- [x] The test plan and commands are included.
- [x] Correctness, performance, and model-evaluation results are included.
- [x] Documentation changes are not required because there is no public API or
  configuration change.

</details>

### CODE REVIEW

**Comment by yzyDavid** [comment]
#49969 is a better optimization, thx to the mention of @benchislett
